ragger_tests/application_client/response_unpacker.py

```diff
@@ -0,0 +1,236 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from enum import IntEnum
+from typing import NamedTuple, Tuple, TypeVar
+from struct import unpack
+from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
+from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePublicKey, ECDSA, SECP256K1
+from cryptography.hazmat.primitives.asymmetric import ec, utils
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives._serialization import Encoding, PublicFormat
+from abc import ABC, abstractmethod
+
+
+PK = TypeVar('PK', bound="IsPublicKey")
+
+class IsPublicKey(ABC):
+    @abstractmethod
+    def verify_signature(self, signature: bytes, hash: bytes) -> bool:
+        pass
+
+    @abstractmethod
+    def serialize(self) -> bytes:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def unpack(cls, raw: bytes) -> PK:
+        pass
+
+@dataclass
+class Curve25519PublicKey(IsPublicKey):
+    wrapped_key: Ed25519PublicKey
+
+    def serialize(self) -> bytes:
+        return self.wrapped_key.public_bytes_raw()
+
+    def verify_signature(self, signature: bytes, hash: bytes) -> bool:
+        try:
+            self.wrapped_key.verify(signature=signature, data=hash)
+            return True
+        except Exception as e:
+            raise ValueError("Invalid signature: ", e)
+        
+    @classmethod
+    def unpack(cls, raw: bytes) -> Curve25519PublicKey:
+        wrapped_key = Ed25519PublicKey.from_public_bytes(data=raw)
+        return Curve25519PublicKey(wrapped_key=wrapped_key)
+        
+
+@dataclass
+class Secp256k1PublicKey(IsPublicKey):
+    wrapped_key: EllipticCurvePublicKey
+
+    def serialize(self) -> bytes:
+        return self.wrapped_key.public_bytes(Encoding.X962, PublicFormat.CompressedPoint)
+
+    def verify_signature(self, signature: bytes, hash: bytes) -> bool:
+        try:
+            self.wrapped_key.verify(
+                signature, 
+                hash, 
+                ec.ECDSA(utils.Prehashed(hashes.SHA256()))
+            )
+            return True
+        except Exception as e:
+            raise ValueError("Invalid signature: ", e)
+        
+    @classmethod
+    def unpack(cls, raw: bytes) -> Secp256k1PublicKey:
+        wrapped_key = ec.EllipticCurvePublicKey.from_encoded_point(curve=SECP256K1(), data=raw)
+        return Secp256k1PublicKey(wrapped_key=wrapped_key)
+
+
+class Version(NamedTuple):
+    major: int
+    minor: int
+    patch: int
+
+@dataclass
+class SignedPayload(ABC):
+    key: PK
```

Comment:
implementing classes `SignedPayloadEd25519` and `SignedPayloadSecp256k1` specifies a concrete type `key`